cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java

```diff
@@ -0,0 +1,264 @@
+    List<InputField> result = new ArrayList<>();
+    for (String field : fields) {
+      ListIterator<String> branchIterator = parents.listIterator(parents.size());
+      boolean foundOrigin = false;
+      while (branchIterator.hasPrevious() && !foundOrigin) {
+        String stage = branchIterator.previous();
+        // check if the current field is output by any one of operation created by current stage
+        List<Operation> operations = stageOperations.get(stage);
+        ListIterator<Operation> operationsIterator = operations.listIterator(operations.size());
+        while (operationsIterator.hasPrevious() && !foundOrigin) {
+          Operation operation = operationsIterator.previous();
+          switch (operation.getType()) {
+            case READ:
+              ReadOperation read = (ReadOperation) operation;
+              if (read.getOutputFields().contains(field)) {
+                result.add(InputField.of(prefixedOperationName(stage, read.getName()), field));
+                foundOrigin = true;
+              }
+              break;
+            case TRANSFORM:
+              TransformOperation transform = (TransformOperation) operation;
+              if (transform.getOutputFields().contains(field)) {
+                result.add(InputField.of(prefixedOperationName(stage, transform.getName()), field));
+                foundOrigin = true;
+              }
+              break;
+            default:
+              break;
+          }
+        }
+      }
+    }
+    Set<String> stageInputs = stageDag.getNodeInputs(currentStage);
+    if (stageInputs.size() > 1) {
+      return replaceOriginsWithMerge(stageInputs, result);
+    }
+    return result;
+  }
+
+  /**
+   * @return return the topologically sorted nodes till the current stage.
+   */
+  private List<String> getParents(String stageName) {
+    List<String> parents = new ArrayList<>();
+    for (String currentStage : topologicalOrder) {
+      if (currentStage.equals(stageName)) {
+        break;
+      }
+      parents.add(currentStage);
+    }
+    return parents;
+  }
+
+  private String createMergeOperationName(Set<String> inputStages) {
+    Set<String> sortedInputs = new TreeSet<>(inputStages);
+    return prefixedOperationName(Joiner.on(SEPARATOR).join(sortedInputs), "merge");
+  }
+
+  /**
+   * Since we first process all operations (convert them into the format required by platform)
+   * and then add the merge operations, we need to replace the origins with newly added merge
+   * operations if the stage has more than one inputs.
+   *
+   * @param stages input stages to the current stage
+   * @param inputFields list of InputField to the operation recorded by current stage
+   * @return updated list of InputFields
+   */
+  private List<InputField> replaceOriginsWithMerge(Set<String> stages, List<InputField> inputFields) {
+    String mergeOperationName = createMergeOperationName(stages);
```

Comment:
is it always guaranteed that ```addMergeOperations()``` will add this merge? Will it have the same set of fields (or a superset) as the list that was computed here?

Reply:
removed this method as its not needed